File: android/Reguerta/app/src/main/java/com/reguerta/user/presentation/access/AuthShellRouting.kt
package com.reguerta.user.presentation.access

enum class AuthShellRoute {
    SPLASH,
    WELCOME,
    LOGIN,
    REGISTER,
    RECOVER_PASSWORD,
    HOME,
}

data class AuthShellState(
    val backStack: List<AuthShellRoute> = listOf(AuthShellRoute.SPLASH),
) {
    val currentRoute: AuthShellRoute
        get() = backStack.last()

    val canGoBack: Boolean
        get() = backStack.size > 1
}

sealed interface AuthShellAction {
    data class SplashCompleted(val isAuthenticated: Boolean) : AuthShellAction

    data object ContinueFromWelcome : AuthShellAction

    data object OpenRegisterFromLogin : AuthShellAction

    data object OpenRecoverFromLogin : AuthShellAction

    data object SessionAuthenticated : AuthShellAction

    data object SignedOut : AuthShellAction

    data object Back : AuthShellAction
}

fun reduceAuthShell(
    state: AuthShellState,
    action: AuthShellAction,
): AuthShellState =
    when (action) {
        is AuthShellAction.SplashCompleted -> {
            state.resetTo(if (action.isAuthenticated) AuthShellRoute.HOME else AuthShellRoute.WELCOME)
        }

        AuthShellAction.ContinueFromWelcome -> state.push(AuthShellRoute.LOGIN)
        AuthShellAction.OpenRegisterFromLogin -> state.push(AuthShellRoute.REGISTER)
        AuthShellAction.OpenRecoverFromLogin -> state.push(AuthShellRoute.RECOVER_PASSWORD)
        AuthShellAction.SessionAuthenticated -> state.resetTo(AuthShellRoute.HOME)
        AuthShellAction.SignedOut -> state.resetTo(AuthShellRoute.WELCOME)
        AuthShellAction.Back -> state.popOrStay()
    }

private fun AuthShellState.push(route: AuthShellRoute): AuthShellState {
    if (currentRoute == route) return this
    return copy(backStack = backStack + route)
}

private fun AuthShellState.resetTo(route: AuthShellRoute): AuthShellState =
    copy(backStack = listOf(route))

private fun AuthShellState.popOrStay(): AuthShellState {
    if (!canGoBack) return this
    return copy(backStack = backStack.dropLast(1))
}

File: ios/Reguerta/Reguerta/Presentation/Access/AuthShellRouting.swift
import Foundation

enum AuthShellRoute: String, Equatable, Hashable, Sendable {
    case splash
    case welcome
    case login
    case register
    case recoverPassword
    case home
}

struct AuthShellState: Equatable, Sendable {
    var backStack: [AuthShellRoute] = [.splash]

    var currentRoute: AuthShellRoute {
        backStack.last ?? .splash
    }

    var canGoBack: Bool {
        backStack.count > 1
    }
}

enum AuthShellAction: Equatable, Sendable {
    case splashCompleted(isAuthenticated: Bool)
    case continueFromWelcome
    case reauthenticate
    case openRegisterFromWelcome
    case openRegisterFromLogin
    case openRecoverFromLogin
    case sessionAuthenticated
    case signedOut
    case back
}

func reduceAuthShell(state: AuthShellState, action: AuthShellAction) -> AuthShellState {
    switch action {
    case .splashCompleted(let isAuthenticated):
        return state.resetTo(isAuthenticated ? .home : .welcome)
    case .continueFromWelcome:
        return state.push(.login)
    case .reauthenticate:
        return state.resetTo([.welcome, .login])
    case .openRegisterFromWelcome:
        return state.push(.register)
    case .openRegisterFromLogin:
        return state.push(.register)
    case .openRecoverFromLogin:
        return state.push(.recoverPassword)
    case .sessionAuthenticated:
        return state.resetTo(.home)
    case .signedOut:
        return state.resetTo(.welcome)
    case .back:
        return state.popOrStay()
    }
}

private extension AuthShellState {
    func push(_ route: AuthShellRoute) -> AuthShellState {
        guard currentRoute != route else { return self }
        return AuthShellState(backStack: backStack + [route])
    }

    func resetTo(_ route: AuthShellRoute) -> AuthShellState {
        AuthShellState(backStack: [route])
    }

    func resetTo(_ routes: [AuthShellRoute]) -> AuthShellState {
        AuthShellState(backStack: routes)
    }

    func popOrStay() -> AuthShellState {
        guard canGoBack else { return self }
        return AuthShellState(backStack: Array(backStack.dropLast()))
    }
}
